File: backend/app/utils/date_time_helper.py
"""
Date and Time Helper Utilities for Philippine Standard Time (PST/UTC+8)

This module provides centralized functions for converting UTC datetime objects
to Philippine Standard Time and formatting them for display.
"""
from datetime import datetime, timezone, timedelta
from typing import Union, Optional


# Philippine Timezone (UTC+8)
PH_TZ = timezone(timedelta(hours=8))


def to_philippine_time(dt: Union[datetime, str, None]) -> Optional[datetime]:
    """
    Convert a datetime (UTC or naive) to Philippine Standard Time (UTC+8).
    
    Args:
        dt: datetime object (UTC or naive), ISO string, or None
        
    Returns:
        datetime object in Philippine timezone, or None if input is None/invalid
    """
    if dt is None:
        return None
    
    # Handle string input
    if isinstance(dt, str):
        try:
            # Try parsing ISO format
            dt = datetime.fromisoformat(dt.replace('Z', '+00:00'))
        except ValueError:
            try:
                # Try parsing common formats
                dt = datetime.strptime(dt, '%Y-%m-%d %H:%M:%S')
            except ValueError:
                return None
    
    # Handle naive datetime (assume UTC)
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    
    # Convert to Philippine time
    return dt.astimezone(PH_TZ)


def format_datetime_ph(dt: Union[datetime, str, None], format_str: str = '%b %d, %Y %I:%M %p %Z') -> str:
    """
    Format a datetime to Philippine Standard Time string.
    
    Args:
        dt: datetime object (UTC or naive), ISO string, or None
        format_str: strftime format string (default: '%b %d, %Y %I:%M %p %Z')
        
    Returns:
        Formatted string in Philippine time, or 'N/A' if input is None/invalid
    """
    if dt is None:
        return 'N/A'
    
    ph_dt = to_philippine_time(dt)
    if ph_dt is None:
        return 'N/A'
    
    # Replace %Z with 'PST' or 'PHST' for clarity
    formatted = ph_dt.strftime(format_str)
    if '%Z' in format_str:
        formatted = formatted.replace('+08:00', 'PST').replace('UTC+08:00', 'PST')
    
    return formatted


def format_datetime_ph_iso(dt: Union[datetime, str, None]) -> Optional[str]:
    """
    Convert datetime to Philippine Standard Time ISO format string.
    
    Args:
        dt: datetime object (UTC or naive), ISO string, or None
        
    Returns:
        ISO format string in Philippine timezone, or None if input is None/invalid
    """
    ph_dt = to_philippine_time(dt)
    if ph_dt is None:
        return None
    
    return ph_dt.isoformat()


def format_date_ph(dt: Union[datetime, str, None], format_str: str = '%b %d, %Y') -> str:
    """
    Format a datetime to Philippine Standard Time date string (no time).
    
    Args:
        dt: datetime object (UTC or naive), ISO string, or None
        format_str: strftime format string (default: '%b %d, %Y')
        
    Returns:
        Formatted date string in Philippine time, or 'N/A' if input is None/invalid
    """
    if dt is None:
        return 'N/A'
    
    ph_dt = to_philippine_time(dt)
    if ph_dt is None:
        return 'N/A'
    
    return ph_dt.strftime(format_str)


def format_time_ph(dt: Union[datetime, str, None], format_str: str = '%I:%M %p') -> str:
    """
    Format a datetime to Philippine Standard Time time string (no date).
    
    Args:
        dt: datetime object (UTC or naive), ISO string, or None
        format_str: strftime format string (default: '%I:%M %p')
        
    Returns:
        Formatted time string in Philippine time, or 'N/A' if input is None/invalid
    """
    if dt is None:
        return 'N/A'
    
    ph_dt = to_philippine_time(dt)
    if ph_dt is None:
        return 'N/A'
    
    return ph_dt.strftime(format_str)


def get_philippine_now() -> datetime:
    """
    Get current datetime in Philippine Standard Time.
    
    Returns:
        Current datetime in Philippine timezone
    """
    return datetime.now(PH_TZ)


def get_utc_now() -> datetime:
    """
    Get current datetime in UTC (for database storage).
    
    Returns:
        Current datetime in UTC timezone
    """
    return datetime.now(timezone.utc)

File: frontend/src/utils/dateTimeHelper.js
/**
 * Date and Time Helper Utilities for Philippine Standard Time (PST/UTC+8)
 * 
 * This module provides centralized functions for converting UTC datetime strings
 * to Philippine Standard Time and formatting them for display.
 */

/**
 * Format a datetime string to Philippine Standard Time
 * @param {string|Date|null} dateString - ISO datetime string, Date object, or null
 * @param {Object} options - Formatting options
 * @returns {string} Formatted datetime string in Philippine time, or 'N/A' if invalid
 */
export function formatDateTimePH(dateString, options = {}) {
  if (!dateString) return 'N/A'
  
  try {
    const date = new Date(dateString)
    
    if (isNaN(date.getTime())) {
      return 'N/A'
    }
    
    const defaultOptions = {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
      hour12: false,
      timeZone: 'Asia/Manila',
      ...options
    }
    
    const datePart = date.toLocaleDateString('en-US', {
      year: defaultOptions.year,
      month: defaultOptions.month,
      day: defaultOptions.day,
      timeZone: 'Asia/Manila'
    })
    
    const timePart = date.toLocaleTimeString('en-US', {
      hour: defaultOptions.hour,
      minute: defaultOptions.minute,
      hour12: defaultOptions.hour12,
      timeZone: 'Asia/Manila'
    })
    
    return `${datePart} ${timePart}`
  } catch (error) {
    console.error('Error formatting datetime:', error)
    return 'N/A'
  }
}

/**
 * Format a datetime string to Philippine Standard Time (short format)
 * @param {string|Date|null} dateString - ISO datetime string, Date object, or null
 * @returns {string} Formatted datetime string (e.g., "Nov 22, 2025 05:47")
 */
export function formatDateTimeShortPH(dateString) {
  if (!dateString) return 'N/A'
  
  try {
    const date = new Date(dateString)
    
    if (isNaN(date.getTime())) {
      return 'N/A'
    }
    
    const datePart = date.toLocaleDateString('en-US', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      timeZone: 'Asia/Manila'
    })
    
    const timePart = date.toLocaleTimeString('en-US', {
      hour: '2-digit',
      minute: '2-digit',
      hour12: false,
      timeZone: 'Asia/Manila'
    })
    
    return `${datePart} ${timePart}`
  } catch (error) {
    console.error('Error formatting datetime:', error)
    return 'N/A'
  }
}

/**
 * Format a datetime string to Philippine Standard Time (12-hour format with AM/PM)
 * @param {string|Date|null} dateString - ISO datetime string, Date object, or null
 * @returns {string} Formatted datetime string (e.g., "Nov 22, 2025 05:47 AM")
 */
export function formatDateTime12HourPH(dateString) {
  if (!dateString) return 'N/A'
  
  try {
    const date = new Date(dateString)
    
    if (isNaN(date.getTime())) {
      return 'N/A'
    }
    
    const datePart = date.toLocaleDateString('en-US', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      timeZone: 'Asia/Manila'
    })
    
    const timePart = date.toLocaleTimeString('en-US', {
      hour: '2-digit',
      minute: '2-digit',
      hour12: true,
      timeZone: 'Asia/Manila'
    })
    
    return `${datePart} ${timePart}`
  } catch (error) {
    console.error('Error formatting datetime:', error)
    return 'N/A'
  }
}

/**
 * Format only the date part in Philippine Standard Time
 * @param {string|Date|null} dateString - ISO datetime string, Date object, or null
 * @returns {string} Formatted date string (e.g., "Nov 22, 2025")
 */
export function formatDatePH(dateString) {
  if (!dateString) return 'N/A'
  
  try {
    const date = new Date(dateString)
    
    if (isNaN(date.getTime())) {
      return 'N/A'
    }
    
    return date.toLocaleDateString('en-US', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      timeZone: 'Asia/Manila'
    })
  } catch (error) {
    console.error('Error formatting date:', error)
    return 'N/A'
  }
}

/**
 * Format only the time part in Philippine Standard Time
 * @param {string|Date|null} dateString - ISO datetime string, Date object, or null
 * @param {boolean} use12Hour - Whether to use 12-hour format (default: false)
 * @returns {string} Formatted time string (e.g., "05:47" or "05:47 AM")
 */
export function formatTimePH(dateString, use12Hour = false) {
  if (!dateString) return 'N/A'
  
  try {
    const date = new Date(dateString)
    
    if (isNaN(date.getTime())) {
      return 'N/A'
    }
    
    return date.toLocaleTimeString('en-US', {
      hour: '2-digit',
      minute: '2-digit',
      hour12: use12Hour,
      timeZone: 'Asia/Manila'
    })
  } catch (error) {
    console.error('Error formatting time:', error)
    return 'N/A'
  }
}

/**
 * Get current datetime in Philippine Standard Time as ISO string
 * @returns {string} Current datetime in Philippine timezone (ISO format)
 */
export function getPhilippineNowISO() {
  const now = new Date()
  // Convert to Philippine time and return ISO string
  const phTime = new Date(now.toLocaleString('en-US', { timeZone: 'Asia/Manila' }))
  return phTime.toISOString()
}

